katdal/test/test_datasources.py

```diff
@@ -0,0 +1,113 @@
+    n_corrprods = x.shape[2] if x.ndim >= 3 else x.shape[1] / 8
+    chunk_size = 4 * n_corrprods * 8
+    chunks = generate_chunks(x.shape, x.dtype, chunk_size,
+                             dims_to_split=(0, 1), power_of_two=True)
+    return da.from_array(x, chunks)
+
+
+def put_fake_dataset(store, base_name, shape):
+    """Write a fake dataset into the chunk store."""
+    data = {'correlator_data': ramp(shape, dtype=np.float32) * (1 - 1j),
+            'flags': np.zeros(shape, dtype=np.uint8),
+            'weights': ramp(shape, slope=256. / np.prod(shape), dtype=np.uint8),
+            'weights_channel': ramp(shape[:-1], dtype=np.float32)}
+    ddata = {k: to_dask_array(array) for k, array in data.items()}
+    chunk_info = {k: {'chunks': darray.chunks, 'dtype': darray.dtype,
+                      'shape': darray.shape} for k, darray in ddata.items()}
+    push = [store.put_dask_array(store.join(base_name, k), darray)
+            for k, darray in ddata.items()]
+    da.compute(push)
+    return data, chunk_info
+
+
+class TestChunkStoreVisFlagsWeights(object):
+    """Test the :class:`ChunkStoreVisFlagsWeights` dataset store."""
+
+    @classmethod
+    def setup_class(cls):
+        cls.tempdir = tempfile.mkdtemp()
+
+    @classmethod
+    def teardown_class(cls):
+        shutil.rmtree(cls.tempdir)
+
+    def test_construction(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb1'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        vfw = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+        weights = data['weights'] * data['weights_channel'][..., np.newaxis]
+        # Check that data is as expected when accessed via VisFlagsWeights
+        assert_equal(vfw.shape, data['correlator_data'].shape)
+        assert_array_equal(vfw.vis.compute(), data['correlator_data'])
+        assert_array_equal(vfw.flags.compute(), data['flags'])
+        assert_array_equal(vfw.weights.compute(), weights)
+
+    def test_missing_chunks(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb2'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        # Delete a random chunk in each array of the dataset
+        missing_chunks = {}
+        for array, info in chunk_info.items():
+            array_name = store.join(base_name, array)
+            slices = da.core.slices_from_chunks(info['chunks'])
+            index = np.random.randint(len(slices))
+            missing_chunks[array] = slices[index]
+            chunk_name, shape = store.chunk_metadata(array_name, slices[index])
+            os.remove(os.path.join(store.path, chunk_name) + '.npy')
+        vfw = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+        # Check that missing chunks have been replaced by zeroes
```

Comment:
Should maybe check that nothing else has been replaced by zeros; and perhaps offset the ramps so that none of the data is itself zero.

Reply:
Done.